ASM Mitschriften 23.04.21 (#22)
importPI19fromDHGE/dhge-pi19-sem4: ASM-GUENTHER/README.md

	./$(OUTFILE_CPP)
```

# Assembler

## Inline Assembler in C

Über die `asm`-Funktion können Assembler-Befehle direkt in C genutzt werden.

> 

Reviewer: ```suggestion

Für GCC ist [hier](https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html) eine Dokumentation verfügbar
```

```C
int main(){
	asm("mov rax, 7"); // setze den Wert des Register "rax" auf 7
	asm("mov rbx, 35"); // setze den Wert des Register "rbx" auf 35
	asm("add rbx, rax"); // addiere die Register "rax" und "rbx" -> Ergebnis in "rbx"
	return 0;
}
```

Die Werte in den Registern können mit Hilfe eines Debuggers, wie `gdb` betrachtet werden.

### Einfaches Debugging mit `gdb`

1. Mit `gdb` das zu debuggende compilierte Programm öffnen: `gdb .\main.o`
2. Breakpoint setzen: `break linenumber`
3. Programm auführen: `run [argv]`
4. Programm wird an Breakpoint gestoppt

**Zustand des Programmes inspizieren**

- Wert der Register ausgeben: `info registers`
- Stepping: `step` nächste Instruktion (auch in Funktionen springen)
- Stepping: `next` nächste Instruktion (nicht in Funktionen springen)

### Geteilte Variablen zwischen C und Assembler

```C
long a = 7;
long b = 35;
long c;

asm("mov rax, %1;"
		"mov rbx, %2;"
		"add rbx, rax;"
		"mov %0, rbx;"
		: "=r" (c) /* output operands */
		: "r" (a), "r" (b) /* input operands */
		: "rbx" /* list of clobbered registers */
);
```

> - Ein-/Ausgabeoperanden werden durch die Syntax `"contraint" ( operand )` definiert

Reviewer: ```suggestion
- Ein-/Ausgabeoperanden werden durch die Syntax `"constraint" ( operand )` definiert
```

Kleiner Typo

> 	- der `constraint` gibt an, in welchem Register `gcc` die Operanden speichern soll (`r` steht dabei für ein beliebiges der Register)

Reviewer: Da nicht irgendein Register gewählt werden kann, da ja manche für anderen Kram reserviert sind, würde ich lieber schreiben:

```suggestion
	- der `constraint` gibt an, in welchem Register `gcc` die Operanden speichern soll (`r` steht dabei für ein automatisch gewähltes Register)
```

um eindeutig zu sein. Quelle: [hier](http://www.ibiblio.org/gferg/ldp/GCC-Inline-Assembly-HOWTO.html#s5)

	- für Ausgabe-Operanden wird vor dem `constraint` ein `=` gesetzt (z.B. `"=r"`)
	- gibt es auschließlich Eingabe-Operanden, wird dies durch `::` definiert
	- `operand` gibt an, aus/in welchem Wert die Ein-/Ausgabe gelesen/geschrieben werden soll
	- in den `asm`-Befehlen werden die Operanden durch `%index` verwendet

> - die `clobbered registers` geben lediglich an, welche Register von den `asm`-Befehlen verwendet werden

Reviewer: ```suggestion
- die `clobbered registers` geben lediglich an, welche Register von den `asm`-Befehlen schreibend verwendet werden, damit GCC nicht annimmt, dass diese Register am Ende des Assemblerteils noch dieselben Werte haben
```

Author: Gut zu wissen, Günther ist bei uns da nicht so genau drauf eingegangen 😅

> Für GCC ist [hier](https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html) eine Dokumentation verfügbar

```C
int main(){
	asm("mov rax, 7"); // setze den Wert des Register "rax" auf 7
	asm("mov rbx, 35"); // setze den Wert des Register "rbx" auf 35
	asm("add rbx, rax"); // addiere die Register "rax" und "rbx" -> Ergebnis in "rbx"
	return 0;
}
```

Die Werte in den Registern können mit Hilfe eines Debuggers, wie `gdb` betrachtet werden.

### Einfaches Debugging mit `gdb`

1. Mit `gdb` das zu debuggende compilierte Programm öffnen: `gdb ./main.o`
2. Breakpoint setzen: `break linenumber`
3. Programm auführen: `run [argv]`
4. Programm wird an Breakpoint gestoppt

**Zustand des Programmes inspizieren**

- Wert der Register ausgeben: `info registers`
- Stepping: `step` nächste Instruktion (auch in Funktionen springen)
- Stepping: `next` nächste Instruktion (nicht in Funktionen springen)

### Geteilte Variablen zwischen C und Assembler

```C
long a = 7;
long b = 35;
long c;

asm("mov rax, %1;"
		"mov rbx, %2;"
		"add rbx, rax;"
		"mov %0, rbx;"
		: "=r" (c) /* output operands */
		: "r" (a), "r" (b) /* input operands */

> 		: "rbx" /* list of clobbered registers */

Reviewer 2: ```suggestion
		: "rbx", "rax" /* list of clobbered registers */
```

best practice ist meines wissens, alle verwendeten register als clobbered zu markieren.

Author: Gut zu wissen, Günther hat leider nicht viel dazu gesagt. Musste für einen Großteil der Stichpunkte hier nochmal selbst recherchieren.
